src/ifcgeom/taxonomy.h

```diff
@@ -125,6 +125,41 @@ struct item {
 	uint32_t identity() const { return identity_; }
 };
 
+struct implicit_item : public item {
+	DECLARE_PTR(implicit_item)
+	
+	virtual item::ptr evaluate() const = 0;
+};
+
+struct piecewise_function : public item {
+	DECLARE_PTR(piecewise_function)
+
+	std::vector<std::pair<double, std::function<Eigen::VectorXd(double u)>>> spans;
```

Comment:
I see what you mean. That's also what I meant earlier with my comment. Hiding this composition in a lambda is felt less then ideal.

https://github.com/aothms/IfcOpenShell/blob/622a21cd41c4b93b5939be30adc325ad06c9e86a/src/ifcgeom/mapping/IfcGradientCurve.cpp#L49-L56

Instead of `tuple<double,function,function, function>`. How do you feel about:

```C++

// let's try to come up with better names
enum curve_semantics {
  xy,
  z,
  cant
};

std::map<curve_semantics, std::function<Eigen::VectorXd(double u)>>
```

then I would also propose not to merge this with piecewise_function but create another implicit_item subtype.

Reply:
> How do you feel about:

I think it is a good solution. Maybe "horizontal" and "vertical" to use the same terminology from the ifc specification.

The CTs for alignment, excluding the 4th CT-Segments, have horizontal, horizontal+vertical, horizontal+vertical+cant, so there is a hierarchy. std::map doesn't enforce the hierarchy. Is there a data structure that could enforce the hierarchy?

Perhaps horizontal can be a class member and vertical and cant be std::optional?

Thought this doesn't enforce that there must be a vertical if there is cant.

Reply:
Oh thanks for clarifying. In that case. Why don't we just return an Eigen::MatrixXd? which can potentially be just a column vector, doesn't need to be square. Or could be a 4x3 (or 4x4) representing rotation and translation.

Naively it seems more efficient to have a single function instead of three separate ones. I can imagine the code paths for these three functions would be rather similar. It's a matter of returning the 1st derivative in addition to the 0th basically (?).

I can imagine you are considering cases where you calculate data that is not needed. E.g for graphic display the rotation is not needed. What we can also do is pass the shape of the matrix we expect as an argument to the function. (or a bool to differentiate between two cases)

`std::function<Eigen::MatrixXd(double u, const std::array<size_t>& dims)>`

(I guess default arguments is not possible on a std::function)

Reply:
I think an Eigen::MatrixXd would work fine. I only suggested 3-functions because that's the way my literal brain sees the solution, I don't have a lot of experience with the Eigen library, so it doesn't enter my thought process - I'll try to be better in that regard.

In general, there isn't a lot of overhead in computing extra parameters you don't need (z-axis and ref-direction) but this might not be the case as the number of linear placements scales up. Additionally, the z-axis and ref-direction only need to be computed if not provided for by the linear placement.

Reply:
Ok. let's go with MatrixXd then. When we merge your PR #3826 I'll try to add it in.